### PR TITLE
`Expr` variants for boolean operations

### LINKED ISSUE #3268: Introduce `Expr` variants for `IS [NOT] TRUE / FALSE / UNKNOWN`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
In https://github.com/apache/arrow-datafusion/pull/3252 we added support for SQL `IS [NOT] TRUE/FALSE` by mapping to the existing logical binary operator `Is[Not]DistinctFrom`. This resulted in a small PR compared to introducing new `Expr` variants such as `IsNotTrue`, but the downside is that we are now unable to implement type validation/coercion to ensure that the left expression is a boolean value.

**Describe the solution you'd like**
Introduce new `Expr` variants but still map to the physical plan for `IsDistinctFrom`.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by sarahyurick**
# Which issue does this PR close?
Closes #3268

 # Rationale for this change
Hi @andygrove, let me know if this is what you were thinking.

### CODE REVIEW

File: datafusion/proto/src/to_proto.rs [hunk 1]
```diff
@@ -619,6 +619,54 @@ impl TryFrom<&Expr> for protobuf::LogicalExprNode {
                     expr_type: Some(ExprType::IsNotNullExpr(expr)),
                 }
             }
+            Expr::IsTrue(expr) => {
+                let expr = Box::new(protobuf::IsTrue {
+                    expr: Some(Box::new(expr.as_ref().try_into()?)),
+                });
+                Self {
+                    expr_type: Some(ExprType::IsTrueExpr(expr)),
```
Comment (sarahyurick) [inline datafusion/proto/src/to_proto.rs:627]: I haven't been able to properly import datafusion-proto, so I might need help figuring out this section.

**Comment by andygrove** [comment]
Thanks @sarahyurick this looks good so far. I created a PR against your branch to fix the protobuf part.

https://github.com/sarahyurick/arrow-datafusion/pull/1/files

**Comment by sarahyurick** [comment]
Awesome, thanks!

**Comment by andygrove** [comment]
The final piece here is to update the SQL query planner (datafusion/sql/src/planner.rs) so that we map the SQL expressions to these new Logical Plan expressions instead of to `IsDistinctFrom` as in:

```
            SQLExpr::IsTrue(expr) => Ok(Expr::BinaryExpr {
                left: Box::new(self.sql_expr_to_logical_expr(*expr, schema, ctes)?),
                op: Operator::IsNotDistinctFrom,
                right: Box::new(lit(true)),
            }),
```

**Comment by andygrove** [comment]
Actually, one other thing will be to map these new `Expr` to the same physical plan as `Is[Not]DistinctFrom` in the physical planner:

```rust
  Operator::IsDistinctFrom => {
                match (left_data_type, right_data_type) {
                    // exchange lhs and rhs when lhs is Null, since `binary_array_op` is
                    // always try to down cast array according to $LEFT expression.
                    (DataType::Null, _) => {
                        binary_array_op!(right, left, is_distinct_from)
                    }
                    _ => binary_array_op!(left, right, is_distinct_from),
                }
            }
```

**Comment by andygrove** [comment]
On seconid thoughts, let's handle the physical plan side of this as a follow on PR. I filed https://github.com/apache/arrow-datafusion/issues/3288 to track this.

**Review by andygrove — APPROVED** [review]
Thanks @sarahyurick. Let's go ahead and merge this and then we can follow up with integrating with the SQL query planner and the physical plan.

**Pull request merged**